### PR TITLE
plugins/fleet_detector: detect request/response sizes

### PR DESCRIPTION
**Pull request by nmoutschen**
Add new metrics to report on subgraph fetches, and request/response sizes for both supergraph and subgraph operations.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugins/fleet_detector.rs [hunk 1]
```diff
@@ -232,6 +241,128 @@ impl PluginPrivate for FleetDetector {
...
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        if !self.enabled {
+            return service;
+        }
+
+        service
+            // Count the number of request bytes from clients to the router
+            .map_request(move |req: router::Request| router::Request {
+                router_request: req.router_request.map(move |body| {
+                    router::Body::wrap_stream(body.map(move |res| {
```
Comment (bnjjj) [inline apollo-router/src/plugins/fleet_detector.rs:254]: We probably just need https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.inspect

File: apollo-router/src/plugins/fleet_detector.rs [hunk 2]
```diff
@@ -232,6 +241,128 @@ impl PluginPrivate for FleetDetector {
...
+                        })
+                    }))
+                }),
+                context: req.context,
+            })
+            // Count the number of response bytes from the router to clients
+            .map_response(move |res: router::Response| router::Response {
+                response: res.response.map(move |body| {
+                    router::Body::wrap_stream(body.map(move |res| {
+                        res.map(move |bytes| {
```
Comment (bnjjj) [inline apollo-router/src/plugins/fleet_detector.rs:271]: https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.inspect

File: apollo-router/src/plugins/fleet_detector.rs [hunk 3]
```diff
@@ -232,6 +241,128 @@ impl PluginPrivate for FleetDetector {
...
+        service
+            // Count the number of bytes per subgraph fetch request
+            .map_request(move |req: HttpRequest| {
+                let sn = sn_req.clone();
+                HttpRequest {
+                    http_request: req.http_request.map(move |body| {
+                        let sn = sn.clone();
+                        RouterBody::wrap_stream(body.map(move |res| {
+                            let sn = sn.clone();
+                            res.map(move |bytes| {
```
Comment (bnjjj) [inline apollo-router/src/plugins/fleet_detector.rs:305]: https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.inspect

File: apollo-router/src/plugins/fleet_detector.rs [hunk 4]
```diff
@@ -232,6 +241,128 @@ impl PluginPrivate for FleetDetector {
...
+                        })
+                    }
+                    Err(err) => {
+                        // On fetch errors, report the status code as 0
+                        u64_counter!(
+                            "apollo.router.operations.fetch",
+                            "Number of subgraph fetches",
+                            1u64,
+                            subgraph.service.name = sn.to_string(),
+                            http.response.status_code = 0i64
```
Comment (bnjjj) [inline apollo-router/src/plugins/fleet_detector.rs:358]: Should we just mark it as an error without the status_code ?

**Review by bnjjj — COMMENTED** [review]
It needs tests to make sur it works as expected

**Review by jonathanrainer — APPROVED** [review]
LGTM

**Review by loshz — APPROVED** [review]

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Needs tests.

File: apollo-router/src/plugins/fleet_detector.rs [hunk 5]
```diff
@@ -473,3 +469,270 @@ fn get_deployment_type() -> &'static str {
...
+            let _data = hyper::body::to_bytes(http_client_response.http_response.into_body())
+                .await
+                .unwrap();
+
+            // THEN fetch metrics should exist
+            assert_counter!(
+                "apollo.router.operations.fetch",
+                1,
+                &[
+                    KeyValue::new("subgraph.service.name", "subgraph"),
```
Comment (bnjjj) [inline apollo-router/src/plugins/fleet_detector.rs:719]: Nit: can we make sure to use `subgraph.name` instead to be consistent with what we have in the [official subgraph attributes](https://www.apollographql.com/docs/graphos/reference/router/telemetry/instrumentation/standard-attributes#subgraph)

**Review by bnjjj — CHANGES_REQUESTED** [review]

**Review by bnjjj — APPROVED** [review]

**Pull request merged**